cmd/medsky/bgs/bgs.go

```diff
@@ -0,0 +1,1440 @@
+	}))
+
+	if !bgs.ssl {
+		e.Use(middleware.LoggerWithConfig(middleware.LoggerConfig{
+			Format: "method=${method}, uri=${uri}, status=${status} latency=${latency_human}\n",
+		}))
+	} else {
+		e.Use(middleware.LoggerWithConfig(middleware.DefaultLoggerConfig))
+	}
+
+	// React uses a virtual router, so we need to serve the index.html for all
+	// routes that aren't otherwise handled or in the /assets directory.
+	e.File("/dash", "public/index.html")
+	e.File("/dash/*", "public/index.html")
+	e.Static("/assets", "public/assets")
+
+	e.Use(MetricsMiddleware)
+
+	e.HTTPErrorHandler = func(err error, ctx echo.Context) {
+		switch err := err.(type) {
+		case *echo.HTTPError:
+			if err2 := ctx.JSON(err.Code, map[string]any{
+				"error": err.Message,
+			}); err2 != nil {
+				bgs.log.Error("Failed to write http error", "err", err2)
+			}
+		default:
+			sendHeader := true
+			if ctx.Path() == "/xrpc/com.atproto.sync.subscribeRepos" {
+				sendHeader = false
+			}
+
+			bgs.log.Warn("HANDLER ERROR: (%s) %s", ctx.Path(), err)
+
+			if strings.HasPrefix(ctx.Path(), "/admin/") {
+				ctx.JSON(500, map[string]any{
+					"error": err.Error(),
+				})
+				return
+			}
+
+			if sendHeader {
+				ctx.Response().WriteHeader(500)
+			}
+		}
+	}
+
+	// TODO: this API is temporary until we formalize what we want here
+
+	e.GET("/xrpc/com.atproto.sync.subscribeRepos", bgs.EventsHandler)
+	e.GET("/xrpc/com.atproto.sync.requestCrawl", bgs.HandleComAtprotoSyncRequestCrawl)
+	e.POST("/xrpc/com.atproto.sync.requestCrawl", bgs.HandleComAtprotoSyncRequestCrawl)
+	e.GET("/xrpc/com.atproto.sync.listRepos", bgs.HandleComAtprotoSyncListRepos)
+	e.GET("/xrpc/com.atproto.sync.getLatestCommit", bgs.HandleComAtprotoSyncGetLatestCommit)
+	e.GET("/xrpc/_health", bgs.HandleHealthCheck)
+	e.GET("/_health", bgs.HandleHealthCheck)
+	e.GET("/", bgs.HandleHomeMessage)
+
+	admin := e.Group("/admin", bgs.checkAdminAuth)
+
+	// Slurper-related Admin API
+	admin.GET("/subs/getUpstreamConns", bgs.handleAdminGetUpstreamConns)
+	admin.GET("/subs/getEnabled", bgs.handleAdminGetSubsEnabled)
+	admin.GET("/subs/perDayLimit", bgs.handleAdminGetNewPDSPerDayRateLimit)
+	admin.POST("/subs/setEnabled", bgs.handleAdminSetSubsEnabled)
+	admin.POST("/subs/killUpstream", bgs.handleAdminKillUpstreamConn)
+	admin.POST("/subs/setPerDayLimit", bgs.handleAdminSetNewPDSPerDayRateLimit)
+
+	// Domain-related Admin API
+	admin.GET("/subs/listDomainBans", bgs.handleAdminListDomainBans)
+	admin.POST("/subs/banDomain", bgs.handleAdminBanDomain)
+	admin.POST("/subs/unbanDomain", bgs.handleAdminUnbanDomain)
+
+	// Repo-related Admin API
+	admin.POST("/repo/takeDown", bgs.handleAdminTakeDownRepo)
+	admin.POST("/repo/reverseTakedown", bgs.handleAdminReverseTakedown)
+	admin.GET("/repo/takedowns", bgs.handleAdminListRepoTakeDowns)
+
+	// PDS-related Admin API
+	admin.POST("/pds/requestCrawl", bgs.handleAdminRequestCrawl)
+	admin.GET("/pds/list", bgs.handleListPDSs)
```

Comment:
"listing PDS instances" in particular i'd like to make a public Lexicon endpoint. I can take on the task of specifying it.